### PR TITLE
Implement role connections API endpoints and objects

### PR DESCRIPTION
**Pull request by MazeXP**
This PR implements the new application role connections endpoints and objects as documented in https://github.com/discord/discord-api-docs/pull/5668

### CODE REVIEW

File: Backend/Remora.Discord.Caching/API/CachingDiscordRestUserAPI.cs [hunk 1]
```diff
@@ -244,4 +244,64 @@ public async Task<Result<IUser>> GetCurrentUserAsync(CancellationToken ct = defa
...
+        {
+            return cacheResult;
+        }
+
+        var getUserApplicationRoleConnection = await _actual.GetUserApplicationRoleConnectionAsync
+        (
+            applicationID,
+            ct
+        );
+        if (!getUserApplicationRoleConnection.IsDefined(out var userApplicationRoleConnection))
```
Comment (VelvetThePanda) [inline Backend/Remora.Discord.Caching/API/CachingDiscordRestUserAPI.cs:268]: Spacing

**Review by VelvetThePanda — APPROVED** [review]
LGTM

File: Backend/Remora.Discord.Caching/API/CachingDiscordRestUserAPI.cs (hunk 1 above)
Comment (MazeXP) [reply Backend/Remora.Discord.Caching/API/CachingDiscordRestUserAPI.cs:268]: It looks a bit odd for sure, but I think it is fine to not add a line break to keep the logical connection between those 2 lines like above.(https://github.com/Remora/Remora.Discord/blob/main/Backend/Remora.Discord.Caching/API/CachingDiscordRestUserAPI.cs#L205-L206)

**Review by Nihlus — CHANGES_REQUESTED** [review]

File: Backend/Remora.Discord.API.Abstractions/API/Rest/IDiscordRestApplicationAPI.cs [hunk 2]
```diff
@@ -331,4 +331,33 @@ Task<Result<IGuildApplicationCommandPermissions>> EditApplicationCommandPermissi
...
+    /// Updates the application role connection metadata records for the given application..
+    /// </summary>
+    /// <remarks>
+    /// An application can have a maximum of 5 metadata records.
+    /// </remarks>
+    /// <param name="applicationID">The ID of the bot application.</param>
+    /// <param name="records">The metadata records to overwrite the existing ones.</param>
+    /// <param name="ct">The cancellation token for this operation.</param>
+    /// <returns>An creation result which may or may not have succeeded.</returns>
+    Task<Result<IReadOnlyList<IApplicationRoleConnectionMetadata>>> UpdateApplicationRoleConnectionMetadataRecordsAsync
```
[the remark this answers is not in the document]
Comment (MazeXP) [reply Backend/Remora.Discord.API.Abstractions/API/Rest/IDiscordRestApplicationAPI.cs:357]: To my extent it is actually clear what this method is being used for.
It is like `BulkOverwrite*ApplicationCommands`, but for role connection metadata.

**Pull request merged**